proposals/nnnn-regex-syntax-run-time-construction.md

```diff
@@ -0,0 +1,1062 @@
+  public func index(after i: Int) -> Int
+
+  public func index(before i: Int) -> Int
+
+  public subscript(position: Int) -> Element
+}
+```
+
+We propose adding an API to `Regex<AnyRegexOutput>` and `Regex<AnyRegexOutput>.Match` to cast the output type to a concrete one. A regex match will lazily create a `Substring` on demand, so casting the match itself saves ARC traffic vs extracting and casting the output.
+
+```swift
+extension Regex.Match where Output == AnyRegexOutput {
+  /// Creates a type-erased regex match from an existing match.
+  ///
+  /// Use this initializer to fit a regex match with strongly typed captures into the
+  /// use site of a dynamic regex match, i.e. one that was created from a string.
+  public init<Output>(_ match: Regex<Output>.Match)
+
+  /// Returns a typed match by converting the underlying values to the specified
+  /// types.
+  ///
+  /// - Parameter type: The expected output type.
+  /// - Returns: A match generic over the output type if the underlying values can be converted to the
+  ///   output type. Returns `nil` otherwise.
+  public func `as`<Output>(_ type: Output.Type) -> Regex<Output>.Match?
+}
+
+extension Regex where Output == AnyRegexOutput {
+  /// Creates a type-erased regex from an existing regex.
+  ///
+  /// Use this initializer to fit a regex with strongly typed captures into the
+  /// use site of a dynamic regex, i.e. one that was created from a string.
+  public init<Output>(_ match: Regex<Output>)
+
+  /// Returns a typed regex by converting the underlying types.
+  ///
+  /// - Parameter type: The expected output type.
+  /// - Returns: A regex generic over the output type if the underlying types can be converted.
+  ///   Returns `nil` otherwise.
+  public func `as`<Output>(_ type: Output.Type) -> Regex<Output>?
+}
+```
+
+We propose adding API to query and access captures by name in an existentially typed regex and match:
+
+```swift
+extension Regex where Output == AnyRegexOutput {
+  /// Returns whether a named-capture with `name` is present.
+  public func contains(namedCapture name: String) -> Bool
+}
+extension Regex.Match where Output == AnyRegexOutput {
+  /// If a named-capture with `name` is present, returns its value. Otherwise `nil`.
+  public subscript(_ name: String) -> AnyRegexOutput.Element? { get }
+}
+
+extension AnyRegexOutput {
+  /// If a named-capture with `name` is present, returns its value. Otherwise `nil`.
+  public subscript(_ name: String) -> AnyRegexOutput.Element? { get }
+}
+```
+
+Finally, we propose API for creating a regex containing literal string content. This produces an equivalent regex to a string literal embededd in the result builder DSL. As this is much less common than run-time compilation or an embedded literal in the DSL, it has an explicit argument label.
+
+```swift
+extension Regex {
+  /// Produces a regex that matches `literal` exactly, as though every
+  /// metacharacter in it was escaped.
+  public init(quoting literal: String)
```

Comment:
I don't think we should call this parameter "literal", as it may very likely not be a literal, e.g. `Regex(quoting: str1 + str2)`.

Reply:
It's more like it's quoting it literally, making it into a literal from the regex's perspective. It's more than just `\Q\(str)\E` because it will effectively escape any interior `\E`. Not to be confused with a Swift language literal, which includes strings and... also regexes.

Seemed better than `string` but I'm open to alternatives

Reply:
Feels more like `verbatim`

Reply:
SGTM